lib/chef/knife/opc_user_delete.rb

```diff
@@ -19,27 +19,145 @@
 
 module Opc
   class OpcUserDelete < Chef::Knife
+  class DeleteFailed < StandardError ; end
     category "OPSCODE PRIVATE CHEF ORGANIZATION MANAGEMENT"
-    banner "knife opc user delete USERNAME [-d]"
+    banner "knife opc user delete USERNAME [-d] [-R]"
 
     option :no_disassociate_user,
-    :long => "--no-disassociate-user",
-    :short => "-d",
-    :description => "Don't disassociate the user first"
+      :long => "--no-disassociate-user",
+      :short => "-d",
+      :description => "Don't disassociate the user first"
 
+    option :remove_from_admin_groups,
+      long:  "--remove-from-admin-groups",
+      short:  "-R",
+      description: "If the user is a member of any org admin groups, attempt to remove from those groups. Ignored if --no-disassociate-user is set."
+
+    attr_reader :username
     include Chef::Mixin::RootRestv0
 
+    deps do
+      require 'chef/org'
+      require 'chef/org/group_operations'
+    end
+
     def run
-      username = @name_args[0]
+      @username = @name_args[0]
+      admin_memberships = []
+      unremovable_memberships = []
+
       ui.confirm "Do you want to delete the user #{username}"
-      unless config[:no_disassociate_user]
-         orgs =  root_rest.get("users/#{username}/organizations")
-         org_names = orgs.map {|o| o['organization']['name']}
-         org_names.each do |org|
-            ui.output root_rest.delete("organizations/#{org}/users/#{username}")
-         end
+
+      if config[:no_disassociate_user]
+        delete_user(username)
+        return
+      end
+
+      ui.output("Checking organization memberships...")
+      orgs = org_memberships(username)
+
+      if orgs.length > 0
+        ui.output("Checking admin group memberships for #{orgs.length} org(s).")
+        admin_memberships, unremovable_memberships = admin_group_memberships(orgs, username)
+      end
+
+      if admin_memberships.empty?
+        disassociate_user(orgs, username)
+        delete_user(username)
+      else # Has admin memberships:
+        if config[:remove_from_admin_groups]
+          unless unremovable_memberships.empty?
+            raise DeleteFailed.new(error_cant_remove_admin_membership(username, unremovable_memberships))
+          end
+          remove_from_admin_groups(admin_memberships, username)
+        else
+          raise DeleteFailed.new(error_admin_group_member(username,
+                                                          admin_memberships))
+        end
+        disassociate_user(orgs, username)
+        delete_user(username)
+      end
+    rescue DeleteFailed => e
+      ui.output(e.message)
+    end
+
+    def disassociate_user(orgs, username)
+      orgs.each  {|org| org.dissociate_user(username) }
+    end
+
+    def org_memberships(username)
```

Comment:
Minor style nit, but this might be a bit prettier:

```ruby
def org_memberships(username)
  org_data = root_rest.get("users/#{username}/organizations")
  org_data.map do |o|
    Chef::Org.new(o['organization']['name'])
  end
end